File: src/0101-0200/121 - Best Time to Buy and Sell Stock/best_time_to_buy_and_sell_stock.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let (mut profit, mut min) = (0, prices[0]);

        // loop through the vector, starting at 1
        for i in 1..prices.len() {
            // if the current value is less than the minimum value, set the minimum value to the current value
            if prices[i] < min {
                min = prices[i];
            }
            // else if the current value minus the minimum value is greater than the current profit, set the profit to the current value minus the minimum value
            else if prices[i] - min > profit {
                profit = prices[i] - min;
            }
        }

        profit
    }
}

File: src/0101-0200/121 - Best Time to Buy and Sell Stock/best_time_to_buy_and_sell_stock.go
package main

func maxProfit(prices []int) int {
	profit, min := 0, prices[0]

	for _, price := range prices {
		// if the current price is lower than the min price, update the min price to the current price
		// else if the current price minus the min price is greater than the profit, update the profit to the current price minus the min price
		if price < min {
			min = price
		} else if price-min > profit {
			profit = price - min
		}
	}

	return profit
}
